### PR TITLE
Add new trigger in Rapid7 IntSights plugin

### PR DESCRIPTION
**Pull request by r7-mgorny**
### Description

* Add new trigger: new alert
* MC-685, MC-724

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{"body": {"meta": {}, "output": {"alert_ids": ["8378e3m5d0d4v49197z8di6z", "3773f895537y428233422b99", "6925451z519j7g7429en5wy7", "461709558z07991571031769", "61561181186a05000774ee00", "16772609082t56203037ot33", "7795452k901p29588912al29"****]}, "log": "Connect: Connecting...\nrapid7/Rapid7 IntSights:1.2.0. Step name: new_alert\n"}, "type": "trigger_event", "version": "v1"}
{"body": {"meta": {}, "output": {"alert_ids": ["964369s12j5b3u8943m3m097"]}, "log": ""}, "type": "trigger_event", "version": "v1"}
{"body": {"meta": {}, "output": {"alert_ids": []}, "log": ""}, "type": "trigger_event", "version": "v1"}
{"body": {"meta": {}, "output": {"alert_ids": []}, "log": ""}, "type": "trigger_event", "version": "v1"}
{"body": {"meta": {}, "output": {"alert_ids": []}, "log": ""}, "type": "trigger_event", "version": "v1"}
{"body": {"meta": {}, "output": {"alert_ids": ["643392258w9j2o3375d2gom1"]}, "log": ""}, "type": "trigger_event", "version": "v1"}

```

<summary>
docker run --rm -i rapid7/rapid7_intsights:1.2.0 --debug run < tests/new_alert.json
</summary>
</details>

### Validate

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Colors.mk ../../tools/Makefiles/Helpers.mk
--
[*] Running validators
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpExampleValidator" failed! 
	Cause: Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.

Validator "Version Increment Validator" failed! 
	Cause: Remote origin/master not found.'master' branch name changed, update validator


----
[*] Total time elapsed: 2297.131ms

```

<summary>
make validate
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpExampleValidator" failed! 
	Cause: Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.

Validator "Version Increment Validator" failed! 
	Cause: Remote origin/master not found.'master' branch name changed, update validator


----
[*] Total time elapsed: 22397.569ms

```

<summary>
icon-validate --all .
</summary>
</details>

### CODE REVIEW

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 1]
```diff
@@ -386,7 +386,7 @@ actions:
       is_closed:
         title: Closed Status
         description: Status of the alert, either closed or open
-        type: boolean
+        type: string
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:389]: is there a significance to changing this from boolean type to a string? the is_closed name implies a true or false expected value

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 2]
```diff
@@ -615,3 +615,114 @@ actions:
...
+        description: Status of the alert, either closed or open
+        type: string
+        required: false
+        example: Closed
+        enum:
+          - Closed
+          - Open
+      has_indicators:
+        title: Has Indicators
+        description: Show alerts with IOCs results
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:712]:
```suggestion
        description: Shows alerts with IOC results
```

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 3]
```diff
@@ -615,3 +615,114 @@ actions:
         description: New created alert ID
         type: string
         required: true
+triggers:
+  new_alert:
+    title: New Alert
+    description: Run when a new alert that matches the given criteria is created in Intsights
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:621]:
```suggestion
    description: Runs when a new alert that matches the given criteria is created in IntSights
```

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 4]
```diff
@@ -4,7 +4,7 @@ products: [insightconnect]
 name: rapid7_intsights
 title: Rapid7 IntSights
 description: Rapid7 IntSights is disrupting external threat intelligence with a combination of human and automated collection, intelligent analysis, and strategic threat hunting that turns the clear, deep, and dark webs into an intelligence resource that any company can deploy
-version: 1.0.0
+version: 1.2.0
```
Comment (pmara-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:7]: If possible, merge this branch and the action branch together. That way we only have to have one extra version.
Suggestion: make one branch that has both changes and increases to 2.0.0 (due to the type change below) We will review this PR and #1029  separately since they are broken out and then merge the combined branch that you create so that they are released in one go.

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 1 above)
Comment (pmara-r7) [reply plugins/rapid7_intsights/plugin.spec.yaml:389]: This will have to be a major version bump unfortunately but should go through

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 5]
```diff
@@ -615,3 +615,114 @@ actions:
...
+        example: ["Phishing"]
+      severity:
+        title: Severity
+        description: 'List of alerts severity. Allowed values: High, Medium, Low'
+        type: '[]string'
+        required: false
+        example: ["Low"]
+      source_type:
+        title: Source Type
+        description: 'List of alerts source type. Allowed values: Application Store, Cyber Security Blog, Hacking News, Cyber Crime Forum, Hacktivism Forum, Social Media, Facebook, Twitter, LinkedIn, Google Plus, VK, Vimeo, YouTube, IRC Channel, IOC Block List, Credit Card Black Market, Paste Site, Data Leakage Website, Leaked Database, File Sharing Website, Gray Hat Website, Black Market, WHOIS servers, Company Website, Wikileaks, Pinterest, Tumblr, Instagram, Telegram, Webmail, Malware Analysis, Firehol, VRA, Other'
```
Comment (pmara-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:637]:
```suggestion
        description: List of alert's source type. Allowed values: Application Store, Cyber Security Blog, Hacking News, Cyber Crime Forum, Hacktivism Forum, Social Media, Facebook, Twitter, LinkedIn, Google Plus, VK, Vimeo, YouTube, IRC Channel, IOC Block List, Credit Card Black Market, Paste Site, Data Leakage Website, Leaked Database, File Sharing Website, Gray Hat Website, Black Market, WHOIS servers, Company Website, Wikileaks, Pinterest, Tumblr, Instagram, Telegram, Webmail, Malware Analysis, Firehol, VRA, Other
```

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 6]
```diff
@@ -615,3 +615,114 @@ actions:
...
+        example: ["Application Store"]
+      network_type:
+        title: Network Type
+        description: 'List of network type. Allowed values: ClearWeb, DarkWeb'
+        type: '[]string'
+        required: false
+        example: ["DarkWeb"]
+      matched_asset_value:
+        title: Matched Asset Value
+        description: List of matched asset values
```
Comment (pmara-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:649]: Can we give any more info to the user here? A new user might not know what to input here

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 7]
```diff
@@ -615,3 +615,114 @@ actions:
...
+        description: Start date to fetch from in Unix Millisecond Timestamp
+        type: string
+        required: false
+        example: 0
+      found_date_to:
+        title: Found Date To
+        description: End date to fetch to in Unix Millisecond Timestamp
+        type: string
+        required: false
+        example: 1633047102456
```
Comment (pmara-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:682]: Getting a different description source vs found dates here would be useful. What does "source" mean and what is "found"?

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 8]
```diff
@@ -615,3 +615,114 @@ actions:
...
+        required: false
+        example: Flagged
+        enum:
+          - Flagged
+          - Unflagged
+      is_closed:
+        title: Closed Status
+        description: Status of the alert, either closed or open
+        type: string
+        required: false
```
Comment (pmara-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:705]: Please run a test with an enum value not set.

File: plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py [hunk 9]
```diff
@@ -133,7 +133,11 @@ def takedown_request(self, alert_id: str, target: str) -> dict:
         )
 
     def get_alerts(self, alert_params: AlertParams) -> list:
-        return self.make_request("GET", "public/v1/data/alerts/alerts-list", params=alert_params.to_dict()).json()
+        response = self.make_request("GET", "public/v1/data/alerts/alerts-list", params=alert_params.to_dict())
+        if response.text:
+            return response.json()
```
Comment (pmara-r7) [inline plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py:138]: Just in case, wrap this in a try/catch and throw a pluginexception if we can't decode the JSON. Other APIs sometimes return non-json text, even if not originally thought to.

File: plugins/rapid7_intsights/unit_test/test_new_alert.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    @patch("insightconnect_plugin_runtime.Trigger.send", side_effect=MockTrigger.send)
+    def test_trigger(self, mock_request, ss):
+        self.action.run({Input.IS_CLOSED: "Open"})
+
+    @timeout_pass(error_callback=check_error_empty)
+    @timeout_decorator.timeout(2)
+    @patch("requests.request", side_effect=Util.mock_request)
+    @patch("insightconnect_plugin_runtime.Trigger.send", side_effect=MockTrigger.send)
+    def test_trigger_with_empty(self, mock_request, ss):
+        self.action.run({Input.ALERT_TYPE: ["Phishing"]})
```
Comment (pmara-r7) [inline plugins/rapid7_intsights/unit_test/test_new_alert.py:81]: There are a lot of inputs for this trigger, they could use some more test cases. Some ideas: 

- Multiple values from the a list of possible values
- Case with no inputs at all

**Review by pmara-r7 — CHANGES_REQUESTED** [review]
see comment about merging branches. Contact soarcerers team if needed.

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 4 above)
Comment (mberezin-r7) [reply plugins/rapid7_intsights/plugin.spec.yaml:7]: Agreed, this would be much cleaner than making two immediately subsequent releases.

File: plugins/rapid7_intsights/icon_rapid7_intsights/triggers/new_alert/trigger.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            source_date_to=params.get(Input.SOURCE_DATE_TO),
+            found_date_from=params.get(Input.FOUND_DATE_FROM),
+            found_date_to=params.get(Input.FOUND_DATE_TO),
+            assigned=params.get(Input.ASSIGNED),
+            is_flagged=params.get(Input.IS_FLAGGED),
+            is_closed=params.get(Input.IS_CLOSED),
+            has_ioc=params.get(Input.HAS_INDICATORS),
+        )
+
+        while True:
```
Comment (mberezin-r7) [inline plugins/rapid7_intsights/icon_rapid7_intsights/triggers/new_alert/trigger.py:40]: Should we add some timeout escape safety check here?

**Review by mrinehart-r7 — APPROVED** [review]

**Review by mberezin-r7 — DISMISSED** [review]

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 4 above)
Comment (r7-mgorny) [reply plugins/rapid7_intsights/plugin.spec.yaml]: I merged here branch from PR: #1029, so here should be all changes with both branches

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 6 above)
Comment (r7-mgorny) [reply plugins/rapid7_intsights/plugin.spec.yaml]: I changed here description, but in documentation there is no more information about this field. Could we contact with intsight team and ask which value can be here?

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 7 above)
Comment (r7-mgorny) [reply plugins/rapid7_intsights/plugin.spec.yaml]: I changed here description, but in documentation there is no more information about this field. Could we contact with intsight team and ask which value can be here?

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 8 above)
Comment (r7-mgorny) [reply plugins/rapid7_intsights/plugin.spec.yaml:885]: I run test without any inputs:
```
{"body": {"meta": {}, "output": {"alert_ids": ["4142j8p1m2l8h77088n5eb0h"]}
{"body": {"meta": {}, "output": {"alert_ids": ["4142j8p1m2l8h77088n5eb1h"]}, "log": ""}, "type": "trigger_event", "version": "v1"}
```

File: plugins/rapid7_intsights/icon_rapid7_intsights/triggers/new_alert/trigger.py (hunk 11 above)
Comment (r7-mgorny) [reply plugins/rapid7_intsights/icon_rapid7_intsights/triggers/new_alert/trigger.py:40]: This is a trigger. We should add in triggers some timout safety check?

Comment (pmara-r7) [reply plugins/rapid7_intsights/icon_rapid7_intsights/triggers/new_alert/trigger.py:40]: No, this is normal from a trigger. Going to mark this as resolved.

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 4 above)
Comment (pmara-r7) [reply plugins/rapid7_intsights/plugin.spec.yaml]: Thanks. Tagging this for our later group PR review later: should we make this 1.1.0? Don't want to confuse any customers

File: plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py [hunk 12]
```diff
@@ -133,11 +133,42 @@ def takedown_request(self, alert_id: str, target: str) -> dict:
...
+                return response.json()
+        except json.decoder.JSONDecodeError as e:
+            raise PluginException(preset=PluginException.Preset.INVALID_JSON, data=e)
+
+        return []
 
     def add_manual_alert(self, manual_alert_params: ManualAlertParams) -> str:
         return self.make_request("PUT", "public/v1/data/alerts/add-alert", json_data=manual_alert_params.to_dict()).text
 
+    def get_cve(self, cve_ids: [str]) -> list:
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py:148]: maybe the code here could be commented? was personally a bit hard to follow, but if no one else agrees i understand

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 13]
```diff
@@ -133,6 +133,170 @@ types:
...
+      type: string
+      description: Version start excluding
+      required: false
+    VersionStartIncluding:
+      title: Version start including
+      type: string
+      description: Version Start Including
+      required: false
+  cpe:
+    range:
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:158]: should range here be capitalized for uniformity?

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 14]
```diff
@@ -133,6 +133,170 @@ types:
...
+      title: Severity
+      type: string
+      description: Severity
+      required: false
+    intsightsScore:
+      title: IntSights Score
+      type: integer
+      description: IntSights score
+      required: false
+    cvssScore:
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:250]:
```suggestion
    cvsScore:
```

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 15]
```diff
@@ -331,7 +495,7 @@ actions:
         example: ["DarkWeb"]
       matched_asset_value:
         title: Matched Asset Value
-        description: List of matched asset values
+        description: List of matched asset values, example. domain name, company name, IP
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:498]: maybe:
```suggestion
        description: List of matched asset values (examples: domain name, company name, IP)
```

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 16]
```diff
@@ -615,3 +779,130 @@ actions:
...
+        example: ["Low"]
+      source_type:
+        title: Source Type
+        description: "List of alert's source type. Allowed values: Application Store, Cyber Security Blog, Hacking News, Cyber Crime Forum, Hacktivism Forum, Social Media, Facebook, Twitter, LinkedIn, Google Plus, VK, Vimeo, YouTube, IRC Channel, IOC Block List, Credit Card Black Market, Paste Site, Data Leakage Website, Leaked Database, File Sharing Website, Gray Hat Website, Black Market, WHOIS servers, Company Website, Wikileaks, Pinterest, Tumblr, Instagram, Telegram, Webmail, Malware Analysis, Firehol, VRA, Other"
+        type: '[]string'
+        required: false
+        example: ["Application Store"]
+      network_type:
+        title: Network Type
+        description: 'List of network type. Allowed values: ClearWeb, DarkWeb'
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:823]:
```suggestion
        description: 'List of network types. Allowed values: ClearWeb, DarkWeb'
```

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 17]
```diff
@@ -615,3 +779,130 @@ actions:
...
+        type: '[]string'
+        required: false
+        example: ["DarkWeb"]
+      matched_asset_value:
+        title: Matched Asset Value
+        description: List of matched asset values, example. domain name, company name, IP
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:829]:
```suggestion
        description: 'List of matched asset values. Examples: domain name, company name, IP'
```

**Review by jrose-r7 — CHANGES_REQUESTED** [review]

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 18]
```diff
@@ -133,6 +133,170 @@ types:
...
+      required: false
+    intsightsScore:
+      title: IntSights Score
+      type: integer
+      description: IntSights score
+      required: false
+    cvssScore:
+      title: CVS Score
+      type: float
+      description: CVS score
```
Comment (mrinehart-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:253]:
```suggestion
      title: CVSS Score
      type: float
      description: CVSS score
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

File: plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py (hunk 12 above)
Comment (r7-kszczepanskagorna) [reply plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py:148]: We tried to create this code as reusable, but if it is not clear enough we can simplify it a little bit. Is it ok for you?

**Review by pmara-r7 — DISMISSED** [review]
I've reviewed this before, so did a very quick look over but looks good enough to get over the line- the prev version of plugin has a bug that is breaking so I believe its important to push out

File: plugins/rapid7_intsights/unit_test/util.py [hunk 19]
```diff
@@ -30,6 +30,10 @@ def read_file_to_string(filename):
         with open(filename) as my_file:
             return my_file.read()
 
+    @staticmethod
+    def read_file_to_dict(filename):
+        return json.loads(Util.read_file_to_string(os.path.join(os.path.dirname(os.path.realpath(__file__)), filename)))
+
     @staticmethod
     def mock_request(*args, **kwargs):
         class MockResponse:
```
Comment (jrose-r7) [inline plugins/rapid7_intsights/unit_test/util.py:39]: Is this `class MockResponse` indented too far?

Comment (pmara-r7) [reply plugins/rapid7_intsights/unit_test/util.py:39]: this is allowed- its like an "inner class" type pattern seen in other languages / use cases. It means only this function uses this class. It also means that if there were any other MockResponse class found in an import in this file, this locally defined MockResponse would still be used instead of that imported one. Its a way of ensuring this function will always use this exact MockResponse class

Comment (mrinehart-r7) [reply plugins/rapid7_intsights/unit_test/util.py:39]: Similar to this, you can also define functions within functions @jrose-r7 !

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 20]
```diff
@@ -495,7 +495,7 @@ actions:
         example: ["DarkWeb"]
       matched_asset_value:
         title: Matched Asset Value
-        description: List of matched asset values, example. domain name, company name, IP
+        description: 'List of matched asset values (examples: domain name, company name, IP)'
```
Comment (mrinehart-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:498]:
```suggestion
        description: 'List of matched asset values. Examples: domain name, company name, IP'
```
Making this consistent with line 829

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Comment by r7-mgorny** [comment]
@wjohnstone-r7 I changed score to float, but every time I run IOC by value it return integer in score. Can we get some data to test it?

## Assessment
### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 https://example.com Step name: get_indicator_by_value\n",
    "meta": {},
    "output": {
      "first_seen": "https://example.com",
      "geo_location": "US",
      "last_seen": "https://example.com",
      "last_update": "https://example.com",
      "related_campaigns": [],
      "related_malware": [],
      "related_threat_actors": [],
      "score": 10,
      "severity": "Low",
      "sources": [
        {
          "ConfidenceLevel": 3,
          "Name": "Dshield"
        }
      ],
      "system_tags": [
        "bot"
      ],
      "tags": [],
      "type": "IpAddresses",
      "value": "https://example.com",
      "whitelist": false
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/rapid7_intsights:2.0.0 --debug run < tests/get_indicator_by_value.json
</summary>
</details>

**Comment by r7-mgorny** [comment]
## Assessment
### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 https://example.com Step name: get_indicator_by_value\n",
    "meta": {},
    "output": {
      "first_seen": "https://example.com",
      "last_seen": "https://example.com",
      "last_update": "https://example.com",
      "related_campaigns": [],
      "related_malware": [],
      "related_threat_actors": [],
      "score": 12.150537634408602,
      "severity": "Low",
      "sources": [
        {
          "ConfidenceLevel": 1,
          "Name": "PhishTank"
        },
        {
          "ConfidenceLevel": 1,
          "Name": "Phishstats"
        }
      ],
      "system_tags": [
        "phishing"
      ],
      "tags": [],
      "type": "Urls",
      "value": "https://example.com",
      "whitelist": false
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/rapid7_intsights:2.0.0 run < tests/get_indicator_by_value.json
</summary>
</details>

**Review by mberezin-r7 — DISMISSED** [review]

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 8 above)
Comment (mrinehart-r7) [reply plugins/rapid7_intsights/plugin.spec.yaml:885]: @r7-mgorny while this works in local testing, this will fail in-product if a user omits this input (even though it is an optional input, it treats the "empty" input as an empty string). You will need to add an empty string enum option to this input and any other enum inputs within the plugin.

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 21]
```diff
@@ -615,3 +779,130 @@ actions:
...
+          - Unassigned
+      is_flagged:
+        title: Alert Flag Status
+        description: Show flagged/unflagged alerts
+        type: string
+        required: false
+        example: Flagged
+        enum:
+          - Flagged
+          - Unflagged
```
Comment (mrinehart-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:880]:
```suggestion
        enum:
          - Flagged
          - Unflagged
          - ""
```

File: plugins/rapid7_intsights/plugin.spec.yaml [hunk 22]
```diff
@@ -615,3 +779,130 @@ actions:
...
+        example: 1633047102456
+      assigned:
+        title: Alert Assignment
+        description: Show assigned/unassigned alerts
+        type: string
+        required: false
+        example: Assigned
+        enum:
+          - Assigned
+          - Unassigned
```
Comment (mrinehart-r7) [inline plugins/rapid7_intsights/plugin.spec.yaml:871]:
```suggestion
        enum:
          - Assigned
          - Unassigned
          - ""
```

File: plugins/rapid7_intsights/help.md [hunk 23]
```diff
@@ -633,7 +714,80 @@ Example output:
 
 ### Triggers
 
-_This plugin does not contain any triggers._
+#### New Alert
+
+This trigger will run when a new alert that matches the given criteria is created in Intsights.
```
Comment (mrinehart-r7) [inline plugins/rapid7_intsights/help.md:719]:
```suggestion
This trigger will run when a new alert that matches the given criteria is created in IntSights.
```

File: plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py [hunk 24]
```diff
@@ -133,11 +133,35 @@ def takedown_request(self, alert_id: str, target: str) -> dict:
...
+        for _ in range(0, 9999):
+            if cve_ids:
+                query_params["cveId[]"] = cve_ids
+
+            response_cve_list = self.make_json_request("GET", path, params=query_params)
+            content.extend(response_cve_list.get("content", []))
+
+            query_params["offset"] = response_cve_list.get("nextOffset", "")
+            if not query_params["offset"]:
+                break
```
Comment (mrinehart-r7) [inline plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py:161]: This is just pagination code, right?

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]
All optional enum inputs within the plugin code will need to have an additional empty string `""` enum value added to them to prevent validation issues.

File: plugins/rapid7_intsights/plugin.spec.yaml (hunk 8 above)
Comment (r7-mgorny) [reply plugins/rapid7_intsights/plugin.spec.yaml:885]: @mrinehart-r7 
Maybe we should add "any" or "both" enum instead of empty string in plugin.spec and then treat it as empty?

File: plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py (hunk 24 above)
Comment (r7-mgorny) [reply plugins/rapid7_intsights/icon_rapid7_intsights/util/api.py:161]: Yes.

**Comment by mrinehart-r7** [comment]
@r7-mgorny there are still some enums in the plugin.spec.yaml that are optional and do not have empty values as options - lines 538, 547, 556, 889 specifically, and they will need to be addressed. Additionally the plugin will need to be regenerated after adjusting some of the descriptions and enums in the plugin.spec.yaml file.

**Comment by r7-mgorny** [comment]
Here is output from run alert and trigger when `"assigned", "is_closed" and "is_flagged"` are set to empty string `""`

## Input
```
    "input": {
      "frequency": 15,
      "is_flagged": ""
    },
```

```
    "input": {
      "frequency": 15,
      "assigned": ""
    },
```

```
    "input": {
      "frequency": 15,
      "is_closed": ""
    },
```

## Assessment
### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 https://example.com Step name: get_alerts\n",
    "meta": {},
    "output": {
      "alert_ids": [
        "7224r7l0z7p4p81031v0ra3u",
        "7539e614176b401943272z71",
        "5066809c557g9o2802xx8bi6",
        "396675005h28212171383826"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/rapid7_intsights:2.0.0 run < tests/get_alerts.json
</summary>
</details>

<details>

```
{
  "body": {
    "meta": {},
    "output": {
      "alert_ids": [
        "7224r7l0z7p4p81031v0ra3u",
        "7539e614176b401943272z71",
        "5066809c557g9o2802xx8bi6",
        "396675005h28212171383826"
      ]
    },
    "log": "Connect: Connecting...\nrapid7/Rapid7 IntSights:2.0.0. Step name: new_alert\n"
  },
  "type": "trigger_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/rapid7_intsights:2.0.0 run < tests/new_alert.json
</summary>
</details>

**Review by mrinehart-r7 — APPROVED** [review]

**Review by blaxminarayan-r7 — APPROVED** [review]
